File: leetcode/medium/Combinations.py
from typing import List

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        result = [] # This will store all valid combinations.
        
        # The backtracking function to generate combinations.
        # 'start' is the number from which we start considering elements for the current combination.
        # 'current_combination' is the list holding the combination being built.
        def backtrack(start, current_combination):
            # Base case: if the current combination has 'k' elements, it's a valid combination.
            if len(current_combination) == k:
                result.append(current_combination[:]) # Add a copy of the combination to the result.
                return

            # Optimization: Prune the search space.
            # If the remaining elements (n - start + 1) plus the elements already in
            # current_combination (len(current_combination)) are less than k,
            # then it's impossible to form a combination of size k.
            # This check helps to avoid unnecessary recursive calls.
            if len(current_combination) + (n - start + 1) < k:
                return

            # Iterate from 'start' to 'n' to choose the next element.
            for i in range(start, n + 1):
                current_combination.append(i) # Include the current number 'i'.
                backtrack(i + 1, current_combination) # Recurse with the next number.
                current_combination.pop() # Backtrack: remove the last added number to explore other possibilities.

        # Start the backtracking process from number 1 with an empty combination.
        backtrack(1, [])
        return result
            
            
        

File: leetcode/medium/Combinations.java
import java.util.List;
import java.util.ArrayList;

class Solution {
    
    private int n;
    private int k;

    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        this.n = n;
        this.k = k;
        combine(1, new ArrayList<>(), result);
        return result;
    }

    private void combine(int current, List<Integer> path, List<List<Integer>> result) {
        if (k == path.size()) {
            result.add(new ArrayList<>(path));
            return;
        }

        for (int i = current; i <= n; i++) {
            path.add(i);
            combine(i + 1, path, result); 
            path.remove(path.size() - 1);
        }
    }
}